code/src/java/plugin/function/GetFact.java

```diff
@@ -0,0 +1,245 @@
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ * 
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+package plugin.function;
+
+import java.util.Arrays;
+
+import pcgen.base.formula.base.DependencyManager;
+import pcgen.base.formula.base.EvaluationManager;
+import pcgen.base.formula.base.FormulaManager;
+import pcgen.base.formula.base.FormulaSemantics;
+import pcgen.base.formula.base.Function;
+import pcgen.base.formula.base.LegalScope;
+import pcgen.base.formula.base.VarScoped;
+import pcgen.base.formula.inst.ScopeInstanceFactory;
+import pcgen.base.formula.parse.ASTPCGenSingleWord;
+import pcgen.base.formula.parse.ASTQuotString;
+import pcgen.base.formula.parse.Node;
+import pcgen.base.formula.visitor.DependencyVisitor;
+import pcgen.base.formula.visitor.EvaluateVisitor;
+import pcgen.base.formula.visitor.SemanticsVisitor;
+import pcgen.base.formula.visitor.StaticVisitor;
+import pcgen.base.util.FormatManager;
+import pcgen.cdom.base.CDOMObject;
+import pcgen.cdom.base.Loadable;
+import pcgen.cdom.content.fact.FactDefinition;
+import pcgen.cdom.enumeration.FactKey;
+import pcgen.cdom.formula.ManagerKey;
+import pcgen.rules.context.AbstractReferenceContext;
+import pcgen.util.StringPClassUtil;
+
+public class GetFact implements Function
+{
+
+	private static final Class<CDOMObject> CDOMOBJECT_CLASS = CDOMObject.class;
+	private static final Class<String> STRING_CLASS = String.class;
+
+	@Override
+	public String getFunctionName()
+	{
+		return "getFact";
+	}
+
+	@Override
+	public Boolean isStatic(StaticVisitor visitor, Node[] args)
+	{
+		if (args.length == 3)
+		{
+			return (Boolean) args[1].jjtAccept(visitor, null);
+		}
+		return Boolean.TRUE;
+	}
+
+	@Override
+	public FormatManager<?> allowArgs(SemanticsVisitor visitor, Node[] args,
+		FormulaSemantics semantics)
+	{
+		if (args.length == 3)
+		{
+			return allowThreeArguments(visitor, semantics, args);
+		}
+		else
+		{
+			semantics.setInvalid("Function " + getFunctionName()
+				+ " received incorrect # of arguments, expected: 2-3 got "
+				+ args.length + " " + Arrays.asList(args));
+			return null;
+		}
+	}
+
+	private FormatManager<?> allowThreeArguments(SemanticsVisitor visitor,
+		FormulaSemantics semantics, Node[] args)
+	{
+		//Turn scope node into a scope name
+		Node scopeNode = args[0];
+		if (!(scopeNode instanceof ASTQuotString))
```

Comment:
Right now, no, there is no way to have two functions with the same name but different formats allowed as args.  That would require much deeper changes back in the formula system to change that assumption.  That's probably a useful discussion, but longer term change no matter what.  Second, this is using instanceof in order to have a useful error message, not just crash due to an invalid cast.  Third, we don't want to do a double dispatch because right now ASTQuotString is entirely generated by JavaCC/JJTree and thus it is beneficial to avoid too many modifications to that class (and its peers) so as to ease forward compatibility if for some reason the class wants to be changed by the automated tools.  Lastly, we can't defer to .jjtAccept as would normally be done because we want to strictly enforce a *direct* String, not a String brought in by a variable.  I think that answers all of the code best practices that this is not using :)